channels/rail/server/rail_main.c

```diff
@@ -0,0 +1,1616 @@
+	return TRUE;
+}
+
+RailServerContext* rail_server_context_new(HANDLE vcm)
+{
+	RailServerContext* context;
+	RailServerPrivate* priv;
+	context = (RailServerContext*)calloc(1, sizeof(RailServerContext));
+
+	if (!context)
+	{
+		WLog_ERR(TAG, "calloc failed!");
+		return NULL;
+	}
+
+	context->vcm = vcm;
+	context->Start = rail_server_start;
+	context->Stop = rail_server_stop;
+	context->ServerHandshake = rail_send_server_handshake;
+	context->ServerHandshakeEx = rail_send_server_handshake_ex;
+	context->ServerSysparam = rail_send_server_sysparam;
+	context->ServerLocalMoveSize = rail_send_server_local_move_size;
+	context->ServerMinMaxInfo = rail_send_server_min_max_info;
+	context->ServerTaskbarInfo = rail_send_server_taskbar_info;
+	context->ServerLangbarInfo = rail_send_server_langbar_info;
+	context->ServerExecResult = rail_send_server_exec_result;
+	context->ServerGetAppIdResp = rail_send_server_get_app_id_resp;
+	context->ServerZOrderSync = rail_send_server_z_order_sync;
+	context->ServerCloak = rail_send_server_cloak;
+	context->ServerPowerDisplayRequest = rail_send_server_power_display_request;
+	context->ServerGetAppidRespEx = rail_send_server_get_appid_resp_ex;
+	context->priv = priv = (RailServerPrivate*)calloc(1,
+	                       sizeof(RailServerPrivate));
+
+	if (!priv)
+	{
+		WLog_ERR(TAG, "calloc failed!");
+		goto out_free;
+	}
+
+	/* Create shared input stream */
+	priv->input_stream = Stream_New(NULL, RAIL_PDU_HEADER_LENGTH);
+
+	if (!priv->input_stream)
+	{
+		WLog_ERR(TAG, "Stream_New failed!");
+		goto out_free_priv;
+	}
+
+	return context;
+out_free_priv:
+	free(context->priv);
+out_free:
+	free(context);
+	return NULL;
+}
+
+void rail_server_context_free(RailServerContext* context)
+{
+	if (context->priv)
+		Stream_Free(context->priv->input_stream, TRUE);
+
+	free(context->priv);
+	free(context);
+}
+
+HANDLE rail_server_get_event_handle(RailServerContext* context)
+{
+	return context->priv->channelEvent;
+}
+
+UINT rail_server_handle_messages(RailServerContext* context)
+{
+	UINT status = CHANNEL_RC_OK;
+	DWORD bytesReturned;
+	UINT16 orderType;
+	UINT16 orderLength;
+	RailServerPrivate* priv = context->priv;
+	wStream* s = priv->input_stream;
+
+	/* Read header */
+	if (!WTSVirtualChannelRead(priv->rail_channel, 0, (PCHAR)Stream_Pointer(s),
+	                           RAIL_PDU_HEADER_LENGTH, &bytesReturned))
+	{
+		if (GetLastError() == ERROR_NO_DATA)
+			return ERROR_NO_DATA;
+
+		WLog_ERR(TAG,  "channel connection closed");
+		return ERROR_INTERNAL_ERROR;
+	}
+
+	/* Parse header */
+	if ((status = rail_read_pdu_header(s, &orderType, &orderLength)) != CHANNEL_RC_OK)
+	{
+		WLog_ERR(TAG, "rail_read_pdu_header failed with error %"PRIu32"!", status);
+		return status;
+	}
+
+	WLog_DBG(TAG, "Received %s PDU, length:%"PRIu16"",
+	         RAIL_ORDER_TYPE_STRINGS[((orderType & 0xF0) >> 3) + (orderType & 0x0F)], orderLength);
+
+	switch (orderType)
+	{
+		case TS_RAIL_ORDER_HANDSHAKE:
+			{
```

Comment:
maybe take the block one time back

Reply:
As in just formatting? This is how they decode all channel PDUs